Success markers telemetry

alwx opened the pull request:
**Proposed changes**:
- fix #9830

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

alwx commented:
This PR adds events specified in this comment: https://github.com/RasaHQ/rasa/issues/9830#issuecomment-946483085

usc-m commented on docs/docs/telemetry/events.json:
```diff
@@ -422,6 +423,63 @@
+      ]
+    },
+    "Markers Stats Computed": {
+      "description": "Triggered when marker stats has been computed.",
```
```suggestion
      "description": "Triggered when marker stats have been computed.",
```

usc-m commented on docs/docs/telemetry/events.json:
```diff
@@ -422,6 +423,63 @@
+      ]
+    },
+    "Markers Extracted": {
+      "description": "Triggered when markers has been extracted.",
```
```suggestion
      "description": "Triggered when markers have been extracted.",
```

usc-m commented on rasa/cli/evaluate.py:
```diff
@@ -123,6 +124,10 @@ def _run_markers(
         stats_file: (Optional) Path to write out statistics about the extracted
                     markers.
     """
+    telemetry.track_markers_evaluation_initiated(
```
@ka-bu if this merges before your stats/wrap-up one we'll need to change this

usc-m reviewed (commented):
Not 100% sure if we should be referring to as "Markers Evaluation" - maybe "Evaluation - Success Markers" (@aeshky @ka-bu any opinions?). Otherwise looks good though!

aeshky reviewed (changes requested):
Thanks for the PR @alwx. 
Are we not tracking the number of markers defined? (I can see that it's not in the final list [here](https://github.com/RasaHQ/rasa/issues/9830#issuecomment-946483085)). If not, why did we decide to drop it? It's informative because it tells us whether users have 1-2 KPIs or are abusing the feature by defining dozens of things to track.

aeshky commented:
@usc-m sorry I didn't read your comment until after I finished my review:
> Not 100% sure if we should be referring to as "Markers Evaluation" - maybe "Evaluation - Success Markers"

I made a similar observation but suggested a different name. I like your suggestion: "Evaluation - Success Markers"

alwx commented on docs/docs/telemetry/events.json:
```diff
@@ -422,6 +423,63 @@
+        }
+      },
+      "additionalProperties": false,
+      "required": [
+        "count"
+      ]
```
Yep, that's a mistake, fixed it in the latest commit.

alwx commented on rasa/cli/evaluate.py:
```diff
@@ -123,6 +124,10 @@ def _run_markers(
         stats_file: (Optional) Path to write out statistics about the extracted
                     markers.
     """
+    telemetry.track_markers_evaluation_initiated(
```
the event is now called "Markers Extraction Initiated" (based on one of your suggestions) so I renamed this function to have a similar name

alwx commented:
Re-requesting your review because I changed all the names and fixed the issue with `required` in `events.json`. 

> Are we not tracking the number of markers defined?

I think we can do this if it makes sense. When does this even needs to be tracked?

aeshky commented:
> > Are we not tracking the number of markers defined?

> I think we can do this if it makes sense. When does this event need to be tracked? 

What I mean is the number of markers in the config file. So as soon as the config file is processed we would know the number of markers. I can dig into the code to see where this happens. 

@usc-m anything else you want to track? You [suggested](https://github.com/RasaHQ/rasa/issues/9830#issuecomment-938680028) a measure of marker complexity. It's a good idea, but maybe not as trivial to implement so maybe we can add it later. What do you think?

usc-m commented:
> What I mean is the number of markers in the config file. So as soon as the config file is processed we would know the number of markers. I can dig into the code to see where this happens.

Number of markers in the config file should be a matter of counting the number of sub-markers of the marker returned from `Marker.from_config` in the CLI code I believe. (so `len(markers.sub_markers)` in `_run_markers` in rasa\cli\evaluate.py)

> @usc-m anything else you want to track? You suggested a measure of marker complexity. It's a good idea, but maybe not as trivial to implement so maybe we can add it later. What do you think?

Not strictly necessary but something like maximum depth of nested markers, or how wide they get (maximum number of sub-markers under a marker that isn't top-level). Don't think these would be hard to add but also aren't as necessary - I think we'd get some useful info out about how the feature is used but it's something we could add later perhaps?

aeshky commented:
> Not strictly necessary but something like maximum depth of nested markers, or how wide they get (maximum number of sub-markers under a marker that isn't top-level). Don't think these would be hard to add but also aren't as necessary

It tells us how complex the conditions get, and thus how people are using (or potentially abusing?) Markers. 

> I think we'd get some useful info out about how the feature is used but it's something we could add later perhaps?

Definitely useful but not urgent. Let's just count how many Markers are defined for now 👍

aeshky commented on docs/docs/telemetry/events.json:
```diff
@@ -422,6 +423,66 @@
+        }
+      },
+      "additionalProperties": false,
+      "required": [
+        "strategy",
+        "only_extract",
+        "seed",
+        "count"
```
@usc-m are all of these required? 🤔 or just the top two?

usc-m replied:
From the perspective of telemetry I think it makes sense - we'd want to know if it's being used or not. The types of those are `string` or `null` so if it's not present we'd see an explicit null. Would be good to check with someone who understands the telemetry schema here better (though I think this is only used in the docs to explain what data we send back to users and isn't used to actually validate anything internally)

aeshky commented on docs/docs/telemetry/events.json:
```diff
@@ -422,6 +423,66 @@
+      ]
+    },
+    "Markers Stats Computed": {
+      "description": "Triggered when marker stats has been computed.",
```
```suggestion
      "description": "Triggered when marker statistics have been computed.",
```

aeshky replied:
Let's use the full word in text, and reseve "stats" for code :)

usc-m commented:
OK, I think I've added the extra telemetry. Is there any other changes we need to make? Anything more to collect, names all fine etc.?

usc-m commented on rasa/telemetry.py:
```diff
@@ -1005,3 +1009,64 @@ def track_nlu_model_test(test_data: "TrainingData") -> None:
+        {
+            "strategy": strategy,
+            "only_extract": only_extract,
+            "seed": seed,
```
This will report the actual seed used by the user - do we want the actual seed or do we want to just know if they used a seed? Is this something that's worth changing or are we considering the actual seed value not important enough to be careful about not collecting?

aeshky replied:
We only want to know if the user set a seed. I don't think it counts as private info so maybe it doesn't matter if we collect it. But if we can do true/false that might be better.

ka-bu commented on rasa/cli/evaluate.py:
```diff
@@ -139,6 +147,10 @@ def _run_markers(
         )
 
+    # Subtract one to remove the virtual OR over all markers
+    num_markers = len(markers) - 1
```
this is the total number of all conditions and operators used in all marker configurations that were evaluated -- to get the number of user-defined markers we should
1. get rid of the special case [here](https://github.com/RasaHQ/rasa/blob/main/rasa/core/evaluation/marker_base.py#L450) and always add an "ANY_MARKER"
2. `len(markers.sub_markers)` instead of `len(markers)`

ka-bu replied:
@usc-m , I could help and add that while you work on other comments if you like?

usc-m replied:
Yeah, I suggested that too, also the branching factor (maximum number of children under any marker)

ka-bu commented on rasa/core/evaluation/marker_base.py:
```diff
@@ -608,6 +609,9 @@ def evaluate_trackers(
                 tracker_result = self.evaluate_events(tracker.events)
                 processed_trackers[tracker.sender_id] = tracker_result
+
+        processed_trackers_count = len(processed_trackers)
```
do we really want the number of processed trackers or processed sessions (or both)?

usc-m replied:
I think in this case processed trackers is actually useful because we get it before, as a command line argument (the users intent) and after (what they actually got). Might help to highlight issues in our models of how tracker stores work. Right now we don't collect session info anywhere so I'm not sure what info we could get from it, but we could also just collect it as well if you think it would be useful

ka-bu replied:
I have no clue how people use sender_ids and trackers usually, so no idea if that is better - but I guess it won't hurt if we don't add this now (and maybe later)

ka-bu commented on rasa/cli/evaluate.py:
```diff
@@ -147,9 +147,14 @@ def _run_markers(
-    telemetry.track_markers_parsed_count(num_markers)
+    max_depth = markers.depth() - 1
+    # Find maximum branching of marker
+    branching_factor = max(len(marker) - 1 for marker in markers)
```
`branching_factor = max(len(sub_marker) - 1  for marker in markers.sub_markers for for sub_marker in marker)` to exclude the artificial Or marker?

usc-m replied:
Oh yep, otherwise we can end up with cases where the number of markers (which we already get) is returned twice - good spot

ka-bu reviewed (approved):
LGTM

aeshky reviewed (approved):
Looks good. There are still a couple of unresolved typos. Did you miss them? (one spotted by you even 😄)
